### PR TITLE
[Metrics] Allow for running accumulation computations

### LINKED ISSUE #4641: Remove automatic reset after `compute()` in the Metric's API?
## 🚀 Feature
a) Remove automatic reset after `compute()` in the Metric API.
b) Option to automatic log running averages instead of batch-wise results

### Motivation

Some of the metrics normally used, such as 'Recall', 'Intersection over Union',... etc, make no sense if calculated for a small set of images. For example, in binary classification, a batch in which there is no positive case returns 0. However, running averages are much better in these cases. Lightning metrics are automatically reset after calling `compute()`. Why is this behavior? isn't  better to allow calling `compute()` many times, allowing for example, to compute running averages in between. If at some point a reset is needed the `reset()` method can always be called.

### LINKED ISSUE #5166: How to compute metric in each step without reseting the metric
I want to know a metric during the process of an epoch. That is, I want to know the metric value until now. However, if I call `metric.compute()` in each step, it will reset my metric, which is not expected. I read the source code and didn't find a clever way to do that. So could you tell me how to do that?

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #5166 
Fixes #4641 

Multiple users have requested that it should be able to call `compute()` method of a metric without `reset()` being called. The argument is that this allows for computing accumulated metrics in a running fashion. This PR adds a new flag to all metrics called `auto_reset_on_compute` that as default is `True` (current behavior) but when set to `False` disables the automatic `reset()` call after compute. It is then up to the user to manually call `reset()` when they want the metric states reset.

When this new flag is `False` and the metric is called in ddp mode, we will still reset the metric states on all processes != 0. This is to make sure that we do not over-accumulate the states. 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]
LGTM, just a comment unrelated to this PR - I would prefer to have the Args section in `__init__` to be closer to where it is actually used...

File: pytorch_lightning/metrics/metric.py [hunk 1]
```diff
@@ -238,17 +245,26 @@ def compute(self):  # pylint: disable=E0202
...
                 setattr(self, attr, deepcopy(default))
 
+    def reset(self):
+        """
+        This method automatically resets the metric state variables to their default value.
+        """
+        if self.auto_reset_on_compute or not self._internal_reset:
+            reset_fn = self._reset_to_default
+        else:
+            reset_fn = all_except_rank_zero(self._reset_to_default)
```
Comment (teddykoker) [inline pytorch_lightning/metrics/metric.py:264]: Does this make it not reset on rank 0? Why would we not want to reset on rank 0?

**Comment by teddykoker** [comment]
Another possible solution here is to simply not call `reset()` on `compute()`. Initially when I wrote the code I thought this would make sense as most of the time you would want to call both, but perhaps the verbosity is better.

PyTorch has a similar pattern with `optimizer.step()`, `optimizer.zero_grad()`. Most of the time you want to just call both, but it makes sense to have the user call both manually.

When logging a metric we could still call both in the logger

**Comment by teddykoker** [comment]
We might want to do something like this for `_wrap_compute`:
```python
 def _wrap_compute(self, compute):
        @functools.wraps(compute)
        def wrapped_func(*args, **kwargs):
            # return cached value
            if self._computed is not None:
                return self._computed

            dist_sync_fn = self.dist_sync_fn
            if (dist_sync_fn is None
                    and torch.distributed.is_available()
                    and torch.distributed.is_initialized()):
                # User provided a bool, so we assume DDP if available
                dist_sync_fn = gather_all_tensors
           
            synced = False
            if self._to_sync and dist_sync_fn is not None:              
                # ADDED: cache prior to syncing
                self._cache = {attr: getattr(self, attr) for attr in self._defaults.keys()}
                
                # sync (this was already there)
                self._sync_dist(dist_sync_fn)
                synced = True

            self._computed = compute(*args, **kwargs)
            if synced:
                 # ADDED: if we synced, restore to cache
                for attr, val in self._cache.items():
                    setattr(self, attr, val)
           
            # NO LONGER RESET
            # self.reset()
            return self._computed
     return wrapped_func
```
Note how we cache the state if syncing, and restore after syncing, so that the accumulation will not be messed up.

**Comment by SkafteNicki** [comment]
@teddykoker personally I would prefer that solution. I only proposed this way, such that it did not break the current setup.
Let me close this PR, and then we implement your suggestion in another PR together with a warning in the docs that user should note that prior to v1.2 that compute also calls reset.
